[Sridhar Pooja] iP

wowwwp opened the pull request:
## Introducing Ella

> Organizing is cool


Ella helps you organize all your tasks. It is:
* Easy to use
* **Fast**
* Remembers stuff

To get started:
1. Download the file from [here](https://github.com/wowwwp/ip/releases)
2. Double click it
3. Now its ready to use 🥇 

**Features**:
- [ ] Add events
- [ ] Find tasks by name
- [ ] Mark and unmark tasks
- [ ] Arrange tasks by deadline 😮 

If your a Java Programmer, you can use it to practice Java too. Here is the `main` method:
```
public class Launcher {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

wowwwp committed:
Rename file to Ella

wowwwp committed:
Add greet and exit functionality

wowwwp committed:
Add echo functionality to user input

wowwwp committed:
Update greet and exit messages

wowwwp committed:
Implement task and list command functionality

wowwwp committed:
Modify main function to handle new commands

wowwwp committed:
Implement Task class to represent a task

wowwwp committed:
Modify the output strings of certain methods for standardization

wowwwp committed:
Modify Ella to add in done and undone functionality

wowwwp committed:
Implement Event, Deadline, and ToDo as subclasses of Task

wowwwp committed:
Update .gitignore to ignore .class files

wowwwp committed:
Modify toString() method to standardize output among the child classes

wowwwp committed:
Implement todo, event, deadline command functionalities

wowwwp committed:
Update Ella to change line type and fixed bugs

wowwwp committed:
Standardize output strings in markAsDone() and markAsUndone() methods

wowwwp committed:
Set up files for automated testing of Ella

wowwwp committed:
Custom exception class for commands which does not follow conventions

wowwwp committed:
Refactor Ella to handle exceptions and clean up redundant logic

wowwwp committed:
Update Ella to handle more errors

wowwwp committed:
Add more commands to test Ella

wowwwp committed:
Update error messages in parseEvent and parseDeadline

wowwwp committed:
Update EXPECTED.TXT to show new error messages

wowwwp committed:
Implement delete command functionality in Ella

wowwwp committed:
Update test cases to test delete command

wowwwp committed:
Implement storage class to handle loading and updating of tasks in JSON file

wowwwp committed:
Modify Task class to be abstract and define abstract getDates() method

Task class does not have any methods to retrieve dates from child
classes.

Storage Class needs the date attribute in child classes like Deadline
and Event for the saving and loading tasks correctly.

Modify Task to be an abstract class with an abstract method getDates().
getDates() can be defined in the respective child classes to return the
number of dates required for the class.

Using abstract methods in this case is preferable as to not introduce
unnecessary methods in classes that do not need them. Each child
class returns different number of dates, and this approach ensures
flexibility.

wowwwp committed:
Implement Constructors and getDates() method in Deadline, Event and ToDo

wowwwp committed:
Modify Ella to load tasks and save tasks after each session

wowwwp committed:
Modify .gitignore to exclude data folder

wowwwp committed:
Modify Ella to handle IO exception

wowwwp committed:
Merge branch 'branch-Level-7'

wowwwp committed:
Modify all dates from String to LocalDateTime

wowwwp committed:
Modify Ella to use LocalDateTime instead of String for dates

wowwwp committed:
Add compareDate() method and TaskComparator class for comparing Tasks

Tasks cannot be compared meaningfully.

With LocalDateTime being used instead of Strings, it is essential to
have a consistent way of comparing Tasks.

A new abstract method compareDate() will return the date which we can
use to compare the task. TaskComparator will rank Tasks based on that
date.

This method will ensure consistency as there are tasks having multiple
dates associated with them and some having none.

wowwwp committed:
Add arrangeTasks method to sort tasks by date

wowwwp committed:
Add max date for tasks

wowwwp committed:
Modify arrangeTasks to list only tasks which are unmarked

wowwwp committed:
Merge branch 'branch-Level-8'

wowwwp committed:
Refactor code to introduce UI class

wowwwp committed:
Modify Storage to save tasks.json file correctly

wowwwp committed:
Update tests file to test for new additions

wowwwp committed:
Add abstract class Command and implement Commands

All the commands and functionalities are executed through switch-case
statement in Main.

The Main function is cluttered and as more commands are introduced
it will get longer.

An abstract class Command is defined and all commands will be subclasses
of Command. Functionalities of each command will be encapsulated in
its class.

This approach allows new commands to be added in easily without
cluttering the Main class.

wowwwp committed:
Create Parser class to encapsulate the parsing logic

wowwwp committed:
Implement TaskList class to encapsulate logic related to handling tasks

wowwwp committed:
Modify Storage to use TaskList for updating tasks

wowwwp committed:
Fix minor bugs in Ui class

wowwwp committed:
Refactor Ella to use Command, Ui and Parser class

The logic for parsing, displaying messages to user as well as executing
commands are found in Ella.

Ella is very long and cluttered and does multiple tasks.

Separate the various components into various classes. Ui handles
user input related tasks, Parser will parse the user input and Command
will execute the functionalities of a given command.

This ensures code will be more organized as well as easier to read.

wowwwp committed:
Modify test cases to test more commands

wowwwp committed:
Move functionalities from TaskList to respective classes

wowwwp committed:
Update message for EXPECTED.TXT

wowwwp committed:
Reoragnized functions in Ella

wowwwp committed:
Modify expected output for test case

wowwwp committed:
Merge branch 'branch-A-MoreOOP'

wowwwp committed:
Organize Java files into packages and add corresponding import statement

wowwwp committed:
Modify file path due to packages

wowwwp committed:
Merge branch 'branch-A-Packages'

wowwwp committed:
Merge remote-tracking branch 'origin/add-gradle-support'

wowwwp committed:
Add GSON dependency to build.gradle file

wowwwp committed:
Merge branch 'branch-add-gradle'

wowwwp committed:
Modify config in build.gradle

wowwwp committed:
Merge checkValidity and checkUserInput functions to remove redundancy

wowwwp committed:
Add JUnit tests for DeadlineCommand, EventCommand and ParserTest

wowwwp committed:
Merge branch 'branch-A-JUnit'

wowwwp committed:
Modify mainClass in build.gradle

wowwwp committed:
Merge branch 'branch-A-Jar'

wowwwp committed:
Rearrange import statements

wowwwp committed:
Add JavaDoc comments to files in command module

wowwwp committed:
Add JavaDoc comments to the files in the task module

wowwwp committed:
Add JavaDoc comments to the files in the utils module

wowwwp committed:
Modify files in Command module to adhere to coding standard

wowwwp committed:
Modify files in task module to adhere to coding standard

wowwwp committed:
Modify files in utils module to adhere to coding standard

wowwwp committed:
Modify files in Test to adhere to coding standards

wowwwp committed:
Modify Ella to adhere to coding standards

wowwwp committed:
Add JavaDoc comments for Ella

wowwwp committed:
Implement Find command to find tasks matching keyword

wowwwp committed:
Add find command logic into Parser

wowwwp committed:
Merge branch 'branch-A-JavaDoc'

wowwwp committed:
Resolved conflicts caused by merging master and branch-A-CodingStandard

wowwwp committed:
Resolve conflicts from merge and add new command

wowwwp committed:
Minor formatting edits

mogmyij commented on src/main/java/Ella/Ella.java:
```diff
@@ -0,0 +1,76 @@
+package Ella;
```
maybe the package name should not contain capital letters, per the coding standard.

mogmyij commented on src/main/java/Ella/task/Deadline.java:
```diff
@@ -0,0 +1,53 @@
+ */
+
+public class Deadline extends Task {
+    private final LocalDateTime by;
```
maybe constant variables should be named in all caps

mogmyij commented on src/main/java/Ella/task/TaskComparator.java:
```diff
@@ -0,0 +1,21 @@
+ * order (earliest deadline first). If two tasks have the same deadline, their order
+ * relative to each other remains unchanged.</p>
+ */
+public class TaskComparator implements Comparator<Task> {
```
any reason why not have the Task class implement the comparable interface which could allow the use of ArrayList methods such as .sort(null) which will make use of the class' natural ordering?

mogmyij reviewed (commented):
Looks good.
just some small changes and questions

hzk-lab commented on src/main/java/Ella/command/ArrangeCommand.java:
```diff
@@ -0,0 +1,60 @@
+     *
+     * @param tasks An ArrayList containing Task Objects
+     */
+    public void arrangeTasks(ArrayList<Task> tasks) {
```
Good work in all your use of Java Stream processing!

hzk-lab commented on src/main/java/Ella/utils/Storage.java:
```diff
@@ -0,0 +1,134 @@
+package Ella.utils;
```
Excellent use of Gson for JSON operations!

hzk-lab commented on src/main/java/Ella/utils/Parser.java:
```diff
@@ -0,0 +1,175 @@
+     * @return An array containing the task description and a date.
+     * @throws IndexOutOfBoundsException If the user did not provide a task and a date.
+     */
+    public static String[] parseDeadline(String[] splits) throws IndexOutOfBoundsException {
```
The parseDeadline method correctly enforces the /by separator, but consider trimming whitespace from the split parts to prevent unexpected issues with extra spaces. This also applies top other methods that handling users' command.

hzk-lab commented on src/main/java/Ella/utils/TaskList.java:
```diff
@@ -0,0 +1,67 @@
+package Ella.utils;
```
Nice job providing detailed JavaDoc comments! The descriptions of the methods and parameters are clear and informative, making the code more maintainable.

wowwwp committed:
Add initial UI files

wowwwp committed:
Add UI configuration files for layout and styling

wowwwp committed:
Refactor files to return Strings for UI functionality

The output statements are directly printed from within the functions.

To display the messages in the user interface, output statements
must be returned as Strings.

Modify the output of all the commands and helper functions to return
Strings.

wowwwp committed:
Merge branch 'Level-10'

wowwwp committed:
Add assert statements to code

There are no assert statements checking assumptions which should hold
true.

There may be bugs introduced due to underlying assumption.

Assert statements are used to validate assumption as they are a good
way to catch inconsistencies.

wowwwp committed:
Rename package Ella to ella to follow coding standard

wowwwp committed:
Refactor code to improve code quality

There are methods which are too long, variables which are not properly
named and unnecessary variables.

Changes:
* Breaking up long methods
* Updating method and variable names
* Removing unnecessary import statements
* Make happy path prominent

This is to ensure code is readable as well as maintainable.

wowwwp committed:
Merge pull request #1 from wowwwp/A-Assertions

Add assert statements to code

wowwwp committed:
Resolve merge conflict

wowwwp committed:
Merge pull request #2 from wowwwp/A-CodeQuality

A-code quality

wowwwp committed:
Make minor edits to ArrangeCommand

wowwwp committed:
Merge branch 'C-Sort'

wowwwp committed:
Modify code to adhere to coding style

wowwwp committed:
Merge branch 'A-Checkstyle'

wowwwp committed:
Set up CI

wowwwp committed:
Fix DateTimeFormatter to ensure consistent formatting

wowwwp committed:
Remove unused imports

wowwwp committed:
Merge branch 'A-CI'

wowwwp committed:
Replace Ella and User display images

wowwwp committed:
Update layout and design of Ella GUI

wowwwp committed:
Update README.md

wowwwp committed:
Merge branch 'A-BetterGui'

wowwwp committed:
Add check for task status before mark/unmark commands

wowwwp committed:
Fix phrasing in output messages for clarity

wowwwp committed:
Add support for additional date formats

wowwwp committed:
Update README.md with instructions for adding tasks in JSON file

wowwwp committed:
Add Ui image

wowwwp committed:
Update README.md with instructions to add tasks in JSON file

wowwwp committed:
Fix minor issues in files
